src/build_helper/src/compiletest.rs

```diff
@@ -0,0 +1,130 @@
+//! Types representing arguments to compiletest.
+
+use std::fmt;
+use std::str::FromStr;
+
+pub use self::Mode::*;
+
+macro_rules! string_enum {
+    ($(#[$meta:meta])* $vis:vis enum $name:ident { $($variant:ident => $repr:expr,)* }) => {
+        $(#[$meta])*
+        $vis enum $name {
+            $($variant,)*
+        }
+
+        impl $name {
+            $vis const VARIANTS: &'static [Self] = &[$(Self::$variant,)*];
+            $vis const STR_VARIANTS: &'static [&'static str] = &[$(Self::$variant.to_str(),)*];
+
+            $vis const fn to_str(&self) -> &'static str {
+                match self {
+                    $(Self::$variant => $repr,)*
+                }
+            }
+        }
+
+        impl fmt::Display for $name {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                fmt::Display::fmt(self.to_str(), f)
+            }
+        }
+
+        impl FromStr for $name {
+            type Err = String;
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                match s {
+                    $($repr => Ok(Self::$variant),)*
+                    _ => Err(format!(concat!("unknown `", stringify!($name), "` variant: `{}`"), s)),
+                }
+            }
+        }
+    }
+}
+
+// Make the macro visible outside of this module, for tests.
+#[cfg(test)]
+pub(crate) use string_enum;
+
+string_enum! {
+    #[derive(Clone, Copy, PartialEq, Eq, Debug, Hash)]
+    pub enum Mode {
```

Comment:
Please call this TestMode

Reply:
why? the command line flag is called `--mode`, no? so why shouldn't the struct match?

Reply:
Because build_helpers is a shared dep between bootstrap, compiletest, and some other bootstrap tools. bootstrap *itself* already has a [`Mode`](https://github.com/rust-lang/rust/blob/6067b36314ab5eb2eb47cecc464545ba58e1ad24/src/bootstrap/src/lib.rs#L204-L209), not to mention `Mode` is a very generic term.

Reply:
i would think its position in the `compiletest` module would be adequate context.

i can if you really want me to, it's just gonna take a bit since it's refactoring across several crates.

Reply:
Yes. I would prefer if this is called `TestMode` and not just `Mode`.